### PR TITLE
Fix StringView's array ctor, and align ZStringView.

### PR DESCRIPTION
**Pull request by BillyONeal**
* Previously, the const char&(arr)[Sz] overloads of StringView and ZStringView were outcompeted by their const char* overloads, so they had no effect. This makes the "const char*" overload a template as well, so that the array version can engage.
* Also aligns / orders StringView and ZStringView functions such that they line up between each other for easier diffing, since functionality provided by one should probably be provided by the other as well.
* Also moves functionality out of the header for ZStringView consistent with StringView.

### CODE REVIEW

**Review by ras0219-msft — APPROVED** [review]
LGTM, I think it would also make sense to have ZStringView derive StringView to reduce code duplication

**Comment by BillyONeal** [comment]
Given the edge case in https://github.com/microsoft/vcpkg-tool/pull/469/commits/0d57060633541d694812941336d2dd26059db624 I think we should consider just outright deleting the array ctor and saying "callers, if you want to avoid strlen you must StringLiteralize".

File: include/vcpkg/base/stringview.h [hunk 1]
```diff
@@ -39,20 +44,23 @@ namespace vcpkg
...
         StringView substr(size_t pos, size_t count = std::numeric_limits<size_t>::max()) const noexcept;
 
-        constexpr char operator[](size_t pos) const noexcept { return m_ptr[pos]; }
-
     private:
         const char* m_ptr = 0;
         size_t m_size = 0;
     };
 
+    // intentionally not hidden friends
```
Comment (strega-nil-ms) [inline include/vcpkg/base/stringview.h:63]: why are these intentionally not hidden?

**Review by strega-nil-ms — APPROVED** [review]

File: include/vcpkg/base/stringview.h (hunk 1 above)
Comment (BillyONeal) [reply include/vcpkg/base/stringview.h:63]: Because we want

```
StringView sv{"hello"};
sv == "world"
```

to compile.

Comment (BillyONeal) [reply include/vcpkg/base/stringview.h:63]: Hmmm come to think of it that may actually work while still hidden friends. I'll just delete the comment.

**Comment by ras0219-msft** [comment]
Can we do constexpr strlen on string literals?

**Comment by BillyONeal** [comment]
> Given the dirent edge case, I'd prefer the resolution "if you want to not use strlen, use StringLiteral", especially since `strlen` on tiny strings is cheap.

Yeah I've convinced myself of that. It's also the status quo.

> Can we do constexpr strlen on string literals? I think if we can rely on it being constexpr, that would resolve all the issues.

strlen is not constexpr in C++20; I don't know if it is in 23

**Comment by BillyONeal** [comment]
> Is that because of language deficiencies or because standards?

Yes. It's difficult in real implementations where those things are often asm blobs and thus can't be marked constexpr, and it's difficult in standards because it's difficult to touch things owned by WG14

**Comment by BillyONeal** [comment]
Resolution in the other direction over at https://github.com/microsoft/vcpkg-tool/pull/478

**Pull request closed without merging**